arbor/backends/gpu/matrix_fine.cu

```diff
@@ -45,43 +45,39 @@ void scatter(const T* from, T* to, const I* p, unsigned n) {
 template <typename T, typename I>
 __global__
 void assemble_matrix_fine(
-        T* d,
-        T* rhs,
-        const T* invariant_d,
-        const T* voltage,
-        const T* current,
-        const T* conductivity,
-        const T* cv_capacitance,
-        const T* area,
-        const I* cv_to_cell,
-        const T* dt_intdom,
-        const I* cell_to_intdom,
-        const I* perm,
+        T* __restrict__ d,
```

Comment:
We have these `__restrict__` pointers here, and then "pre-load" the values all the same in the loops.

I would remove these `__restrict__` keywords, and write the loops in such a way that the compiler can be certain that there is no aliasing.

Reply:
As https://github.com/arbor-sim/arbor/pull/1026 has been merged, it might be better to go the other route of keeping the restrict.

Reply:
I am not a big fan of relying on `__restrict__` when better written versions of kernels (many of which I wrote!) would remove the need for them. In this PR we are adding `__restrict__` and also rewriting loops to avoid reading a value twice that could have been aliased: we should pick one approach and stick to it.

The other big danger of `__restrict__` is that we use it when it is not valid, and create all manner of nasty aliasing bugs.

Reply:
Hi,

`const __restrict__` in particular does a bit more for us here, it also enables use of constant memory (quoting a conversation with NVIDIA here). But it only takes effect if all pointers are qualified as `__restrict__`. Also, #1026 has been merged, adding `__restrict__` qualifiers to most kernels. About the invalid use: Most of these kernels are single use, ie there are very few callers and most often only one. That would make abuse pretty obvious and also quite unlikely. I also like that the qualifier comminucates our intend.

That said, there is is only a minor impact, if any and I can remove the qualifiers if needed.

```
|-------+-------|
|  with |   w/o |
|-------+-------|
| 1.996 | 2.005 |
| 2.006 | 2.013 |
| 2.011 | 2.014 |
| 2.014 | 2.014 |
| 2.015 | 2.017 |
| 2.022 | 2.024 |
| 2.024 | 2.024 |
| 2.025 | 2.027 |
| 2.026 | 2.029 |
| 2.042 | 2.046 |
|-------+-------|
```